### ISSUE TITLE
Update Site Cloner - Doesn't work with Block Themes

### ISSUE THREAD

**Issue by outdoor2kode** [WordPress/wordcamp.org#1315]
### Describe the bug

https://wordpress.slack.com/archives/C08M59V3P/p1715272078520129

Details pending completion.

**Comment by pkevan** [comment]
@renintw did you get anywhere with this? It's come up a few times that block themes don't clone easily - it was suggested that the create block theme community plugin could get us someway there.

**Comment by pkevan** [comment]
when attempting to clone a block theme (tt4) noting some things i'm noticing missing:

- remote CSS location
- template parts and templates
- patterns
- content from source site (unsure if this is needed?)

If content/media isn't cloned is it a true clone of another site?

**Comment by outdoor2kode** [comment]
When I previously helped develop the website, and from feedback I saw over time, it doesn't matter if content and media aren't cloned. But patterns, template parts, and templates are important and really what people want. The remote CSS location doesn't need to be cloned either since a new folder should be created each time.

**Comment by outdoor2kode** [comment]
Some findings:

1. The site cloner in the local environment is not functioning.
2. The customizer additional CSS and the site editor’s additional CSS in the local environment cannot be opened. Attempts to fix deprecated functions, upgrade the WordPress version, database, and plugins did not resolve the issue. Currently, both the sandbox and production environments are functioning correctly. It is suspected that the issue stems from the setup process of the local environment.
3. How the current clone works:
    1. The target site switches the theme to the theme of the cloned site.
    2. Editor styles like: 
        a. Variations 
        b. Colors 
        c. Shadows 
        d. Layout 
        e. Block settings 
        f. Additional CSS are not cloned.
    3. The customizer’s Additional CSS is cloned to the target site’s Customizer’s Additional CSS.
    4. Remote CSS (such as specified GitHub CSS file) is also cloned to the target site’s Customizer’s Additional CSS.
    5. The input in Appearance -> Fonts is not cloned.
    6. The input in Appearance -> Remote CSS is not cloned, but as mentioned above, the CSS content is directly cloned to the customizer’s additional CSS.
    7. All Post Types, CPT posts, and media are not cloned.

**Comment by outdoor2kode** [comment]
Testing the "Create Block Theme" plugin:

1. The exported ZIP file does not include the patterns and templates added to the site.
2. The plugin itself does not provide an import feature.

**Comment by StevenDufresne** [comment]
> The exported ZIP file does not include the patterns and templates added to the site.

I'm pretty certain it should.

**Comment by outdoor2kode** [comment]
> I'm pretty certain it should.

I retested this locally rather than on the testing site again. Patterns didn't get exported. As for custom templates, they will be exported with the prefix 'wp-custom-template'.

**Update**
Found that patterns are exported only after the "Save Changes" in the Create Block Theme tool is pressed. Initially, the default settings appear as shown in the image, but they won't take effect until "Save Changes" again. However, if users create patterns without selecting "synced," those patterns still won't be exported.

<img width="562" alt="image" src="https://github.com/user-attachments/assets/f3a2595d-7810-443f-a64b-5f47fe84f6b2">

Additionally, any images added to the site must have the "Localize Images" option selected in order to be exported.

**Comment by renintw** [comment]
Here's an idea for further testing: try `Create Theme` > `Clone Theme` first and see how the cloned theme can be fetched and activated on a new site. This way, all patterns (synced and non-synced), templates, etc., will be retained. This logic can be added to the site cloner, specifically where it clones the old site's theme and activates it on the new site. This would complement the site cloner's existing ability to clone RemoteCSS, making the cloning process more complete.

**Comment by dd32** [comment]
Noting that there's two main aspects to this:
 - When on a Block Theme, you can't clone any sites
 - A Block Theme site can't be cloned
 - The search/Filtering of the list no longer works.

As a FYI to others, this is not likely to be resolved anytime soon.

**Comment by wputasic** [comment]
Root cause of breakages 1 and 3: the cloner registers its Customizer assets only on `admin_enqueue_scripts`, which `customize.php` only fires via the widgets component — and WP 5.9+ doesn't load that component on block themes. So the assets never load and the section renders as a dead shell (which is also why switching to a classic theme "fixes" it).

#1758 registers them on `customize_controls_enqueue_scripts` instead. Verified on WP 6.8.3: with Twenty Twenty-Five active the list, search, preview, and clone flow all work again; classic themes unaffected. Breakage 2 (cloning *from* a block-theme site) is out of scope — that still needs the bigger rework.

Could this ship as a partial fix, so organizers on the default theme can use the cloner again?

**Comment by wputasic** [comment]
# Proposal: bring the Site Cloner to block themes (re: #1315)

Following up on #1315. I dug into why the cloner is broken on block themes and built a working
proof-of-concept for the "full solution" @dd32 flagged. This splits cleanly into two parts.

---

## Part 1 — the dead UI (breakages #1 and #3): a one-hook fix, already up as #1758

The reason the cloner shows nothing on a block theme isn't the Customizer, Jetpack, or the clone
logic — it's a severed asset-enqueue chain:

- The plugin registers its Customizer JS/CSS + `_wcscSettings` (REST URL/nonce) only on
  `admin_enqueue_scripts`.
- `customize.php` never fires that hook directly. On classic themes it fires *indirectly*, via the
  Customizer **widgets** component (`WP_Customize_Widgets::enqueue_scripts()` →
  `do_action('admin_enqueue_scripts','widgets.php')`).
- Since WP 5.9, the widgets component isn't loaded when a **block theme** is active
  (`WP_Customize_Manager::__construct()`), so the assets never register and the "Clone Another
  WordCamp" section renders as a dead shell — no site list, no search, no error.

That's also why switching to a classic theme "fixes" it. **#1758** registers the assets on
`customize_controls_enqueue_scripts` (a hook `customize.php` always fires). Verified on WP 6.8.3:
with Twenty Twenty-Five active, the list, search, preview, and clone flow all work again; classic
themes are unaffected. This restores cloning **from classic-theme sites** regardless of the active
theme — the majority of the cloneable back-catalog.

---

## Part 2 — cloning a block-theme site's design (breakage #2)

### The reframe that makes it tractable

A freshly-activated block theme writes **zero** database rows — all of its design lives in theme
files until an organizer customizes it in the Site Editor, at which point WordPress persists
*override* rows. So a clone doesn't reconstruct a site; it **copies the DB deltas** from source to
target. And since nearly every camp runs the default (Twenty Twenty-Four), source and target usually
share a theme — meaning in the common case there's **no theme switch at all**, just a data copy.

This matches the 2024 findings in this thread (thanks @renintw / @outdoor2kode): content and media
aren't required, but **global styles, templates, template parts, patterns (and fonts)** are "what
people want."

### What must be copied

| Design store | Location | Theme-scoped |
|---|---|---|
| Global styles + **Additional CSS** (block themes store it in `styles.css` inside this post, *not* the classic `custom_css` post) | `wp_global_styles` | yes |
| Templates / template parts | `wp_template` / `wp_template_part` | yes |
| Synced patterns | `wp_block` | no |
| Navigation | `wp_navigation` | no |
| Fonts | the `wc-fonts-options` option (WordCamp doesn't use the core Font Library, so there are no font-file copies to worry about) | n/a |
| Referenced media | `attachment` posts + files | no |

Remote CSS keeps the current behaviour — copy the sanitized content, not the URL (each site gets its
own repo/folder), per @outdoor2kode's note.

### The hard parts — and that they're solved in the PoC

- **Opaque `ref` post-IDs.** `core/navigation` and `core/block` embed integer post IDs that mean
  nothing on the target. The clone copies the referenced posts and rewrites every `ref` in the copied
  templates/parts to the target's new IDs. (Neither core export nor `create-block-theme` does this for
  a DB→DB copy — `create-block-theme` is export-only, and it drops navigation entirely.)
- **Media is blog-local.** Every image is an attachment ID + a blog-scoped URL. The PoC sideloads each
  referenced file into the target's library and remaps IDs, `wp-image-N` classes, and URLs — including
  background images set via a block's `style.background`.
- **Two multisite gotchas** (both bit me, both handled): after `switch_to_blog()`, `wp_upload_dir()`
  gains a bogus `mu/` segment (core runs from `public_html/mu/`) — fixed with wcpt's
  `_fix_wc_upload_dir` filter; and `wp_get_attachment_url()` across a switch returns a different URL
  form than the pretty `/files/` URL in content — fixed by capturing `home_url('/files/'.$relpath)`.

### What the PoC does

A network-internal engine (`wp wcsc-block-clone <source_blog> <target_blog>`) that copies the stores
above with:
- **`--mode=replace`** (default) or **`--mode=merge`** (non-destructive: keep the target's design,
  add only what's missing, preserve local edits).
- **`--theme=switch`** (default: target adopts the source's theme — lossless) or **`--theme=keep`**
  (cross-theme: re-slug the design to the target's own theme).
- Idempotent re-clone (cloned attachments are tagged so they don't accumulate).

---

## Validation — tests and checkpoints

Tested locally (Docker, WP 6.8.3) cloning a customized Twenty Twenty-Four site (fixtures: custom
palette + Additional CSS, an edited template, a header part referencing a nav by `ref`, a synced
pattern referenced by `ref`, an embedded image, a group background image, and a fonts option) into a
second site. Three suites, all green:

**Replace + switch — 22/22.** Target theme switched onto the source's; global styles copied
(`isGlobalStylesUserThemeJSON` flag intact, Additional CSS, custom palette); navigation, synced
pattern, template, template part all present; **nav `ref` remapped** to the target's new nav post;
**pattern `ref` remapped** to the target's new pattern; template-part `theme` attr correct; **embedded
image + group background image** sideloaded, IDs/classes/URLs remapped, files on the target's disk,
no leftover source URLs; fonts option copied. Re-running stays 22/22 (idempotent).

**Merge (non-destructive) — 6/6.** After a replace clone, I injected a local edit into the target's
Additional CSS and deleted its navigation, then ran a merge clone. The local edit was preserved, the
original cloned CSS remained, the missing navigation was re-added, and nothing was duplicated. Report:
`clobbered: skipped`, `created: {navigation: 1}`, `skipped: {global_styles, patterns, templates×2, fonts}`.

**Cross-theme (`--theme=keep`) — 11/11.** Cloned the Twenty Twenty-Four design onto a target running
Twenty Twenty-Five: the target stayed on TT5, global styles / templates / parts were re-tagged to TT5,
the `wp-global-styles-<slug>` name and template-part `theme` attr re-slugged, refs remapped, media
localized. (Structural validity only — see limitations.)

**Checkpoints** captured at each stage of the build (MVP → media → background images + merge →
cross-theme), so any stage is reproducible.

---

## Honest limitations

- **Cross-theme is structurally valid but not visually faithful.** A design references its theme's
  `theme.json` presets; a different target theme may define different ones, so re-slugged designs can
  fall back. True fidelity needs a preset-mapping layer — which is *why* core doesn't offer cross-theme
  migration. Same-theme (`switch`) is lossless, and it's the dominant real-world case.
- The **apply step** is currently a network-internal action (WP-CLI/REST), not the Customizer
  changeset-publish path — the Customizer can't preview/switch block themes. The selection UI can
  stay in the Customizer (works post-#1758); the apply becomes a "Clone now" action. Preview would be
  screenshot-based (as today) rather than live.
- Media dedupe is tag-and-replace, not content-hash. Merge existence is by (type, slug, theme); a
  "skip only if edited since last clone" refinement would use a provenance marker.

---

## Proposed phasing

- **Phase 0 — #1758**: revive the cloner UI on block-theme sites (classic-theme sources work again).
- **Phase 1 — same-theme design clone**: global styles + Additional CSS, templates, parts, synced
  patterns, navigation, fonts, with ref-remap. Covers the dominant TT4→TT4 case.
- **Phase 2 — media localization** and **merge mode** (both prototyped).
- **Phase 3 — cross-theme** (prototyped, structural) and UI/preview polish.

## Decisions I'd like the group's input on

1. Is **#1758** OK to ship as the standalone fix for breakages #1/#3?
2. For the full solution: is **same-theme-only** acceptable for a first release?
3. Confirm **media and content are out of scope** for v1 (thread suggests yes).
4. **Replace vs merge** as the default when re-cloning onto a customized site?
5. Where should the **apply UI** live — keep the Customizer picker + a "Clone now" action, or a
   dedicated Appearance screen?

The engine logic is prototyped and verified locally (test results above), so the remaining work for a
review-ready **Phase-1 PR** is integration, the apply-UI, sanitization/caps, and tests. I'm happy to
open that PR once we've settled the UI approach (decision 5 above); media, merge, and cross-theme
would follow as incremental PRs (all already prototyped). Glad to share the engine or do a
screen-share in the meantime if that's useful.
